### ISSUE TITLE
임베디드 레이어를 사용해야 하는 이유

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/deep-learning-readinglist#147]
https://medium.com/towards-data-science/deep-learning-4-embedding-layers-f9a02d55ac12

**Comment by sweaterr** [comment]
심층 학습에 관한 본 시리즈의 제 4 부에 오신 것을 환영합니다. 처음 3 개 항목과 이 게시물 사이에 약간의 지연이 있음을 눈치 챘을 것입니다. 이 시리즈의 초기 목표는 심층 학습에 대한 fast.ai 과정과 함께 작성하는 것이 었습니다. 그러나 나중에 강의의 개념이 겹치는 경우가 있으므로 먼저 코스를 끝내기로했습니다. 이 방법을 통해 이러한 주제에 대한보다 자세한 개요를 제공 할 수 있습니다. 이 블로그에서는 여러 강의 (4-6)를 포괄하는 개념을 다루려고합니다. 실제로는 실용적으로 유용합니다. Embedding Layers.

소개시 임베딩 레이어의 개념은 상당히 외래 일 수 있습니다. 예를 들어, Keras 문서는 "양의 정수 (인덱스)를 고정 크기의 고밀도 벡터로 변환"이외의 다른 설명을 제공하지 않습니다. 이러한 유형의 문서는 처음으로 팝업 될 것이므로 빠른 Google 검색을 사용하면 많은 도움이 되지 않을 수도 있습니다. 그러나, Keras의 문서는 모든 일이 일어난다는 것을 의미합니다. 그렇다면 왜 임베디드 레이어를 사용해야합니까? 다음은 두 가지 주요 이유입니다.

1. 원 핫 코드 벡터는 고차원 이고 sparse 합니다 . 우리가 자연어 처리 (NLP)를 하고 2000 단어의 사전을 가지고 있다고 가정합시다. 즉, one-hot 인코딩을 사용할 때 각 단어는 2000 개의 정수를 포함하는 벡터로 나타냅니다. 그리고 이 정수 중 1999 개는 0입니다. 큰 데이터 세트에서 이 접근법은 계산 상 효율적이지 않습니다.

2. 각 임베딩 벡터는 신경망을 학습하는 동안 업데이트 됩니다. 이 글의 맨 위에 있는 이미지를 보았다면 다차원 공간에서 단어 사이의 유사점을 발견 할 수 있습니다. 이를 통해 단어 간의 관계뿐만 아니라 임베디드 레이어를 통해 벡터로 바뀔 수  있는 모든 것을 시각화 할 수 있습니다.

이 개념은 여전히 ​​다소 모호 할 수 있습니다. 단어의 예를 사용하여 임베디드 레이어의 기능을 살펴 보겠습니다. 그럼에도 불구하고, embeddings의 기원은 단어 embeddings에서 온다. 더 많이 읽으려는 경우 word2vec를 검색 할 수 있습니다. 이 문장을 예로 들어 봅시다 (심각하게 생각하지 마십시오) :

> “deep learning is very deep”

임베딩 레이어를 사용하는 첫 번째 단계는 이 문장을 인덱스로 인코딩하는 것입니다. 이 경우 각 고유 단어에 색인을 할당합니다. 이 문장은 다음과 같이 보입니다.

> 1 2 3 4 1

임베딩 행렬이 다음에 생성됩니다. 우리는 얼마나 많은 'latent factor'이 각 index 에 배정되는지를 결정합니다. 기본적으로 이것은 벡터 크기를 의미합니다. 일반적인 사용 사례는 32와 50 같은 길이입니다.이 게시물에서  index 당 6 개의 'latent factor' 을 지정하여 가독성을 유지합시다. 임베딩 행렬은 다음과 같습니다.

![](https://cdn-images-1.medium.com/max/800/1*Di85w_0UTc6C3ilk5_LEgg.png

**Comment by sweaterr** [comment]
Not Just Word Embeddings

이 이전 예제는 자연어 처리의 세계에서 단어 삽입이 매우 중요하다는 것을 보여주었습니다. 그들은 우리가 다른 방식으로 포착하기가 어려운 언어로 관계를 포착 할 수있게 해줍니다. 그러나 임베디드 레이어를 사용하여 단순한 단어 이외의 많은 것들을 포함 할 수 있습니다. 현재 진행중인 연구 프로젝트에서는 온라인 사용자 행동을 퍼가기 위해 임베디드 레이어를 사용하고 있습니다. 이 경우 '포털 Y에서 페이지 유형 X의 페이지보기'또는 '스크롤 X 픽셀'과 같은 사용자 동작에 색인을 할당합니다. 이 색인은 일련의 사용자 동작을 구성하는 데 사용됩니다.

전통적인'기계 학습 모델 (SVM, Random Forest, Gradient Boosted Trees)과 심층 학습 모델 (심 신경 망, 반복 신경 망)의 비교에서 이 임베딩 접근법은 깊은 신경 네트워크에서 매우 잘 작동한다는 것을 알게되었습니다.

'전통적인'기계 학습 모델은 피쳐 엔지니어링된 표 형식의 입력에 의존합니다. 이것은 우리가 연구자로서 어떤 것이 특징으로 바뀌는지를 결정한다는 것을 의미합니다. 이러한 경우 기능은 방문한 홈페이지의 양, 완료된 검색의 양, 스크롤 한 총 픽셀 수입니다. 그러나 피쳐 엔지니어링을 할 때 공간 (시간) 차원을 포착하는 것은 매우 어렵습니다. 깊은 학습과 임베드 레이어를 사용함으로써 우리는 일련의 사용자 행동 (지표)을 모델의 입력으로 제공함으로써 이 공간 차원을 효율적으로 포착 할 수 있습니다.

저의 연구에서 Gated Recurrent Unit / Long-Term Term Memory를 가진 Recurrent Neural Network가 가장 잘 수행되었습니다. 결과는 매우 가깝습니다. '전통적인'기능 설계된 모델에서 Gradient Boosted Trees가 가장 잘 수행되었습니다. 이 연구에 대한 블로그 게시물을 앞으로 더 자세히 작성하겠습니다. 나는 내 다음 블로그 포스트가 재발 신경 네트워크를 더 자세히 조사 할 것이라고 생각한다.

다른 연구는 MOOCs (Piech et al., 2016)의 학생 행동과 온라인 패션 상점 (Tamhane et al., 2017)을 통한 사용자 경로를 인코딩하기 위해 레이어를 포함하는 방법을 연구했습니다.

**Comment by sweaterr** [comment]
Recommender Systems

임베딩 레이어는 추천 시스템에서 스파 스 매트릭스 문제를 처리하는 데 사용될 수도 있습니다. 깊은 학습 과정 (fast.ai)은 추천 시스템을 사용하여 임베디드 레이어를 소개하기 때문에 여기서도 탐색하려고합니다.

추천 시스템은 어디에서나 사용되고 있으며, 아마도 당신은 매일 그들에 의해 영향을 받고 있습니다. 가장 일반적인 예로는 Amazon의 제품 권장 사항과 Netflix의 프로그램 권장 시스템이 있습니다. Netflix는 실제로 자신의 추천 시스템에 가장 적합한 공동 필터링 알고리즘을 찾기 위해 1,000,000 달러의 도전을 받았습니다. 이 모델들 중 하나의 시각화를 여기서 볼 수 있습니다.
